src/libServer/Server.cpp

```diff
@@ -193,69 +212,85 @@ Json::Value Server::CreateTransaction(const Json::Value& _json) {
+
+          if (account == nullptr) {
+            throw JsonRpcException(RPC_INVALID_ADDRESS_OR_KEY,
+                                   "To addr is null");
+          }
 
-        else if (!account->isContract()) {
-          throw JsonRpcException(RPC_INVALID_ADDRESS_OR_KEY,
-                                 "Non - contract address called");
-        }
+          else if (!account->isContract()) {
+            throw JsonRpcException(RPC_INVALID_ADDRESS_OR_KEY,
+                                   "Non - contract address called");
+          }
 
-        unsigned int to_shard =
-            Transaction::GetShardIndex(tx.GetToAddr(), num_shards);
-        if (to_shard == shard) {
-          if (!ARCHIVAL_LOOKUP) {
-            m_mediator.m_lookup->AddToTxnShardMap(tx, shard);
-          } else {
-            m_mediator.m_lookup->AddToTxnShardMap(tx, 0);
+          unsigned int to_shard =
+              Transaction::GetShardIndex(tx.GetToAddr(), num_shards);
+          bool sendToDs = false;
+          if (_json.isMember("priority")) {
+            sendToDs = _json["priority"].asBool();
           }
-          ret["Info"] =
-              "Contract Txn, Shards Match of the sender "
-              "and reciever";
-          ret["TranID"] = tx.GetTranID().hex();
-        } else {
-          if (!ARCHIVAL_LOOKUP) {
-            m_mediator.m_lookup->AddToTxnShardMap(tx, num_shards);
+          if ((to_shard == shard) && !sendToDs) {
+            if (!ARCHIVAL_LOOKUP) {
+              m_mediator.m_lookup->AddToTxnShardMap(tx, shard);
+            } else {
+              m_mediator.m_lookup->AddToTxnShardMap(
+                  tx, SEND_TYPE::ARCHIVAL_SEND_SHARD);
+            }
+            ret["Info"] =
+                "Contract Txn, Shards Match of the sender "
+                "and reciever";
+            ret["TranID"] = tx.GetTranID().hex();
           } else {
-            m_mediator.m_lookup->AddToTxnShardMap(tx, 0);
+            if (!ARCHIVAL_LOOKUP) {
+              m_mediator.m_lookup->AddToTxnShardMap(tx, num_shards);
+            } else {
+              m_mediator.m_lookup->AddToTxnShardMap(
+                  tx, SEND_TYPE::ARCHIVAL_SEND_DS);
+            }
+            ret["Info"] = "Contract Txn, Sent To Ds";
+            ret["TranID"] = tx.GetTranID().hex();
           }
-          ret["Info"] = "Contract Txn, Sent To Ds";
-          ret["TranID"] = tx.GetTranID().hex();
-        }
+          return ret;
+        } break;
+
+        case ERROR:
+          throw JsonRpcException(RPC_INVALID_ADDRESS_OR_KEY,
+                                 "Code is empty and To addr is null");
+          break;
+        default:
+          throw JsonRpcException(RPC_MISC_ERROR, "Txn type unexpected");
       }
     } else {
-      LOG_GENERAL(INFO, "No shards yet");
-      throw JsonRpcException(RPC_IN_WARMUP, "Could not create Transaction");
+      throw JsonRpcException(RPC_IN_WARMUP, "No Shards yet");
```

Comment:
consider move this case `num_shards <= 0` before and save some indention for the big chunk of code above.

```
if (num_shards <= 0) {
  throw ...;
}

unsigned int shard = Transaction::GetShardIndex(fromAddr, num_shards);
switch (GetTransactionType(tx)) {
  ...
}
```